x-pack/plugins/security_solution/public/common/components/user_privileges/use_endpoint_privileges.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useEffect, useMemo, useRef, useState } from 'react';
+import { useCurrentUser, useHttp } from '../../lib/kibana';
+import { appRoutesService, CheckPermissionsResponse } from '../../../../../fleet/common';
+
+export interface EndpointPrivileges {
+  loading: boolean;
+  /** If user has permissions to access Fleet */
+  canAccessFleet: boolean;
+  /** If user has permissions to access Endpoint management (includes check to ensure they also have access to fleet) */
+  canAccessEndpointManagement: boolean;
+}
+
+/**
+ * Retrieve the endpoint privileges for the current user.
+ *
+ * **NOTE:** Consider using `usePrivileges().endpointPrivileges` instead of this hook in order
+ * to keep API calls to a minimum.
+ */
+export const useEndpointPrivileges = (): EndpointPrivileges => {
+  const http = useHttp();
+  const user = useCurrentUser();
+  const isMounted = useRef<boolean>(true);
+  const [canAccessFleet, setCanAccessFleet] = useState<boolean>(false);
+  const [fleetCheckDone, setFleetCheckDone] = useState<boolean>(false);
+  const [userRoleCheckDone, setUserRoleCheckDone] = useState<boolean>(false);
+
+  // Check if user can access fleet
+  useEffect(() => {
+    (async () => {
+      try {
+        const fleetPermissionsResponse = await http.get<CheckPermissionsResponse>(
+          appRoutesService.getCheckPermissionsPath()
+        );
+
+        if (isMounted.current) {
+          setCanAccessFleet(fleetPermissionsResponse.success);
+        }
+      } finally {
+        if (isMounted.current) {
+          setFleetCheckDone(true);
+        }
+      }
+    })();
+  }, [http]);
+
+  // Check if user has `superuser` role
+  const isSuperUser = useMemo(() => {
+    if (user?.roles) {
+      setUserRoleCheckDone(true);
```

Comment:
I don't think React allows setting state inside `useMemo`; it will probably complain about this line in runtime. But it looks like you can remove `userRoleCheckDone` altogether without affecting the logic.

Reply:
Not sure I can remove it because the `user` instance is also retrieved async. The `loading` indicator should only be set to `false` when both of these checks have completed, so I think we need to track them both.

I did not see any errors in the console when I ran it.

Reply:
Sorry, what I meant is you can check `user != null` instead of `userRoleCheckDone`. Looks like it should be enough:

```diff
diff --git a/x-pack/plugins/security_solution/public/common/components/user_privileges/use_endpoint_privileges.ts b/x-pack/plugins/security_solution/public/common/components/user_privileges/use_endpoint_privileges.ts
index a419fa552fd..1391aeed01a 100644
--- a/x-pack/plugins/security_solution/public/common/components/user_privileges/use_endpoint_privileges.ts
+++ b/x-pack/plugins/security_solution/public/common/components/user_privileges/use_endpoint_privileges.ts
@@ -29,7 +29,6 @@ export const useEndpointPrivileges = (): EndpointPrivileges => {
   const isMounted = useRef<boolean>(true);
   const [canAccessFleet, setCanAccessFleet] = useState<boolean>(false);
   const [fleetCheckDone, setFleetCheckDone] = useState<boolean>(false);
-  const [userRoleCheckDone, setUserRoleCheckDone] = useState<boolean>(false);
 
   // Check if user can access fleet
   useEffect(() => {
@@ -53,7 +52,6 @@ export const useEndpointPrivileges = (): EndpointPrivileges => {
   // Check if user has `superuser` role
   const isSuperUser = useMemo(() => {
     if (user?.roles) {
-      setUserRoleCheckDone(true);
       return user.roles.includes('superuser');
     }
     return false;
@@ -61,11 +59,11 @@ export const useEndpointPrivileges = (): EndpointPrivileges => {
 
   const privileges = useMemo(() => {
     return {
-      loading: !fleetCheckDone || !userRoleCheckDone,
+      loading: !fleetCheckDone || user != null,
       canAccessFleet,
       canAccessEndpointManagement: canAccessFleet && isSuperUser,
     };
-  }, [canAccessFleet, fleetCheckDone, isSuperUser, userRoleCheckDone]);
+  }, [canAccessFleet, fleetCheckDone, isSuperUser, user]);
 
   // Capture if component is unmounted
   useEffect(
```

But you are right that React does allow setting state inside `useMemo`, I was sure it would complain about setting state during a rendering cycle 🤔